File: Searching/Jump Search/java/JumpSearch.java
import java.util.Scanner;

public class JumpSearch {
    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        System.out.println("Enter the length of Array: ");
        int n = scanner.nextInt();

        System.out.println("Enter the element to be searched:");
        int element = scanner.nextInt();

        int jumpSize = (int) Math.ceil(Math.sqrt(n));
        int[] arr = new int[n];

        System.out.println("Enter array elements: ");
        for(int i=0;i<n;i++){
            arr[i] = scanner.nextInt();
        }

        arr = returnSorredArray(arr);
        int tempIndex = jump(arr, jumpSize, element);

        if(tempIndex > -1){
            int elementIndex = doLinearSearch(arr, tempIndex, jumpSize, element);
            if(elementIndex > -1){
                System.out.println("Element index in sorted array: "+elementIndex);
            }else{
                System.out.println("Element not found!");
            }
        }else{
            System.out.println("Element not found!");
        }

    }

    static int[] returnSorredArray(int[] arr){
        for(int i = 0; i < arr.length - 1; i++){
            for(int j = 0; j < arr.length-i-1; j++){
                if(arr[j+1]<arr[j]){
                    //swap
                    int temp= arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = temp;
                }
            }
        }
        return arr;
    }

    static int jump(int[] arr, int jumpSize, int element){
        for(int i=0; i<arr.length; i++){
            int index = Math.min(jumpSize*i,arr.length-1);
            if(arr[index] >= element){
                return index;

            }
        }
        return -1;
    }

    static int doLinearSearch(int[] arr, int tempIndex, int jumpSize, int element){
        for(int i = tempIndex; i > tempIndex - jumpSize; i--){
            if(arr[i] == element){
                return i;
            }
        }
        return -1;
    }
}

File: Searching/Jump Search/python/jump_search.py
# Author : Shubham Bindal


# Binary Search is better than Jump Search
# but Jump search has an advantage that we traverse back only once
#(Binary Search may require up to O(Log n) jumps
# consider a situation where the element to be search is the smallest element
# or smaller than the smallest).
# So in a systems where jumping back is costly, we use Jump Search.

import math

def jump_search( arr , x , n ): 
      
    block_size = int(math.sqrt(n))
    
    index = 0
    while arr[int(min(block_size, n)-1)] < x: 
        index = block_size 
        block_size += int(math.sqrt(n)) 
        if index >= n: 
            return -1
      
    # perform linear search for x in block beginning with index. 
    while arr[int(index)] < x: 
        index += 1
          
        if index == min(block_size, n): 
            return -1
       
    if arr[int(index)] == x: 
        return index 
      
    return -1

# driver code

# give elements in sorted order
print("Enter the elements of the sorted array in a single line: ")

arr = list(map(int,input().split()))
num = int(input("Enter the number you want to search\n"))

n = len(arr)

(jump_search(arr,num,n))
 
index = jump_search(arr, num, n)

print("Searched_number" , num, "is at index" ,"%.0f"%index) 
